cpp/src/reductions/reductions.cu

```diff
@@ -39,234 +37,155 @@ void gpu_reduction_op(const T *data, const gdf_valid_type *mask,
-        // find needed gridsize
-        // use atmost REDUCTION_BLOCK_SIZE blocks
-        int blocksize = REDUCTION_BLOCK_SIZE;
-        int gridsize = (output_size < REDUCTION_BLOCK_SIZE?
-                        output_size : REDUCTION_BLOCK_SIZE);
-
-        // launch kernel
-        gpu_reduction_op<<<gridsize, blocksize>>>(
-            // inputs
-            data, valid, size,
-            // output
-            output,
-            // action
-            functor,
-            // identity
-            identity,
-            // loader
-            loader
-        );
-    }
-
-};
-
-
-struct DeviceSum {
-    typedef IdentityLoader Loader;
-    typedef DeviceSum second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs + rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
-};
-
-struct DeviceProduct {
-    typedef IdentityLoader Loader;
-    typedef DeviceProduct second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs * rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{1}; }
-};
-
-struct DeviceSumOfSquares {
-    struct Loader {
-        template<typename T>
-        __device__
-        T operator() (const T* ptr, int pos) const {
-            T val = ptr[pos];   // load
-            return val * val;   // squared
-        }
-    };
-    // round 2 just uses the basic sum reduction
-    typedef DeviceSum second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) const {
-        return lhs + rhs;
+    // First thread of each block stores the result.
+    if (tid == 0){
+        genericAtomicOperation(results[0], agg, functor);
     }
+}
 
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
+template<typename T_in, typename T_out, typename Op>
+void ReduceOp(const gdf_column *input,
+                   gdf_scalar* scalar, cudaStream_t stream)
+{
+    T_out identity = Op::template identity<T_out>();
+
+    // allocate temporary memory for the result
+    void *result = NULL;
+    RMM_TRY(RMM_ALLOC(&result, sizeof(T_out), stream));
+
+    // initialize output by identity value
+    CUDA_TRY(cudaMemcpyAsync(result, &identity,
+            sizeof(T_out), cudaMemcpyHostToDevice, stream));
+    CUDA_CHECK_LAST();
+
+    int blocksize = reduction_block_size;
+    int gridsize = (input->size + reduction_block_size -1 )
+        /reduction_block_size;
+
+    // kernel call
+    gpu_reduction_op<<<gridsize, blocksize, 0, stream>>>(
+        static_cast<const T_in*>(input->data), input->valid, input->size,
+        static_cast<T_out*>(result),
+        Op{}, identity, typename Op::Loader{});
+    CUDA_CHECK_LAST();
+
+    // read back the result to host memory
+    // TODO: asynchronous copy
+    CUDA_TRY(cudaMemcpy(&scalar->data, result,
```

Comment:
Agree, since you already do everything else in a stream use `cudaMemcpyAsync` here on `stream`. Then sync the stream after the RMM_FREE.

Reply:
Do you plan to change this to `cudaMemcpyAsync`?